dali/pipeline/operators/decoder/cache/decoder_cache_factory.h

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_DECODER_CACHE_DECODER_CACHE_FACTORY_H_
+#define DALI_PIPELINE_OPERATORS_DECODER_CACHE_DECODER_CACHE_FACTORY_H_
+
+#include <memory>
+#include <string>
+#include <map>
+#include <mutex>
+#include "dali/pipeline/operators/decoder/cache/decoder_cache.h"
+
+namespace dali {
+
+class DLL_PUBLIC DecoderCacheFactory {
+ public:
+  DLL_PUBLIC static inline DecoderCacheFactory& Instance(int device_id) {
```

Comment:
This function shouldn't be inline - it has loads of static data and is DLL-exported. I believe we might run into lots of trouble sooner than later with that. By the way - why do we need to have a factory per-device instead of a cache per-device?

Reply:
Initially it looked better to me to do 
```
DecoderCacheFactory::Instance(device_id).Init(...)
DecoderCacheFactory::Instance(device_id).Get()
...
```
than
```
DecoderCacheFactory::Instance()::Init(device_id, ...)
DecoderCacheFactory::Instance()::Get(device_id)
...
```
but I agree having a factory with multiple caches is more intuitive